## Assignment 2 - Pandas

### Make sure to answer any questions asked and to provide specified outputs

# Part 1: Getting a feel for the data

- Import pandas and numpy
- set random seed to 42 using np.random.seed(42)

In [55]:
import pandas as pd
import numpy as np

np.random.seed(42)

Read in the creatures.csv dataset and assign it to the variable 'creatures' - use read.csv() function

In [56]:
creatures = pd.read_csv("creatures.csv")

Use .info(), .describe(), and .head(10) to get an idea of what the dataset looks like - show outputs for all 3 of these

In [57]:
creatures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 13 columns):
 #   Column      Non-Null Count  Dtype 
---  ------      --------------  ----- 
 0   Num         800 non-null    int64 
 1   Name        800 non-null    object
 2   Type1       800 non-null    object
 3   Type2       414 non-null    object
 4   Total       800 non-null    int64 
 5   HP          800 non-null    int64 
 6   Attack      800 non-null    int64 
 7   Defense     800 non-null    int64 
 8   SpAtk       800 non-null    int64 
 9   SpDef       800 non-null    int64 
 10  Speed       800 non-null    int64 
 11  Generation  800 non-null    int64 
 12  Legendary   800 non-null    bool  
dtypes: bool(1), int64(9), object(3)
memory usage: 75.9+ KB


In [58]:
creatures.describe()

Num      Total          HP      Attack     Defense       SpAtk  \
count  800.000000  800.00000  800.000000  800.000000  800.000000  800.000000   
mean   362.813750  435.10250   69.258750   79.001250   73.842500   72.820000   
std    208.343798  119.96304   25.534669   32.457366   31.183501   32.722294   
min      1.000000  180.00000    1.000000    5.000000    5.000000   10.000000   
25%    184.750000  330.00000   50.000000   55.000000   50.000000   49.750000   
50%    364.500000  450.00000   65.000000   75.000000   70.000000   65.000000   
75%    539.250000  515.00000   80.000000  100.000000   90.000000   95.000000   
max    721.000000  780.00000  255.000000  190.000000  230.000000  194.000000   

            SpDef       Speed  Generation  
count  800.000000  800.000000   800.00000  
mean    71.902500   68.277500     3.32375  
std     27.828916   29.060474     1.66129  
min     20.000000    5.000000     1.00000  
25%     50.000000   45.000000     2.00000  
50%     70.000000   65.000000     3.00000  
75%     90.000000   90.000000     5.00000  
max    230.000000  180.000000     6.00000

In [59]:
creatures.head(10)

Num                       Name  Type1   Type2  Total  HP  Attack  Defense  \
0    1                  Bulbasaur  Grass  Poison    318  45      49       49   
1    2                    Ivysaur  Grass  Poison    405  60      62       63   
2    3                   Venusaur  Grass  Poison    525  80      82       83   
3    3      VenusaurMega Venusaur  Grass  Poison    625  80     100      123   
4    4                 Charmander   Fire     NaN    309  39      52       43   
5    5                 Charmeleon   Fire     NaN    405  58      64       58   
6    6                  Charizard   Fire  Flying    534  78      84       78   
7    6  CharizardMega Charizard X   Fire  Dragon    634  78     130      111   
8    6  CharizardMega Charizard Y   Fire  Flying    634  78     104       78   
9    7                   Squirtle  Water     NaN    314  44      48       65   

   SpAtk  SpDef  Speed  Generation  Legendary  
0     65     65     45           1      False  
1     80     80     60           1      False  
2    100    100     80           1      False  
3    122    120     80           1      False  
4     60     50     65           1      False  
5     80     65     80           1      False  
6    109     85    100           1      False  
7    130     85    100           1      False  
8    159    115    100           1      False  
9     50     64     43           1      False

# Question 1:
  - What datatypes are present within the dataset (Hint: look at .info() output)
    
  - Are there any columns with empty/missing values? (Hint: look at .info() output - columns with no missing values will have 800)
    

1. Boolean, Integers64, and Objects

2. Rows 0 through 2, and Rows 4 through 12

Use .value_counts() on the 'Type1' column and provide the output

In [60]:
type1_counts = creatures['Type1'].value_counts()
print(type1_counts)

Water       112
Normal       98
Grass        70
Bug          69
Psychic      57
Fire         52
Electric     44
Rock         44
Dragon       32
Ground       32
Ghost        32
Dark         31
Poison       28
Steel        27
Fighting     27
Ice          24
Fairy        17
Flying        4
Name: Type1, dtype: int64


# Question 2:
- What type has the most values?

- What type has the least values?


1. Water

2. Flying

# Part 2: Creating a new attribute/column

Create a new column 'CombinedAttack' by adding the 'SpAtk' and 'Attack' column values together across each row

In [61]:
creatures['CombinedAttack'] = creatures['SpAtk'] + creatures['Attack']

Use .describe() again to find the percentile values for this new column

In [62]:
combined_attack_stats = creatures['CombinedAttack'].describe()
print(combined_attack_stats)

count    800.000000
mean     151.821250
std       54.462505
min       20.000000
25%      111.750000
50%      146.500000
75%      182.000000
max      360.000000
Name: CombinedAttack, dtype: float64


Replace the values A, B, and C in the bins below with the 25%, 50%, and 75% values you found using the .describe() value outputs for the new column

- creatures["CombinedAttack"] = pd.cut(creatures["CombinedAttack"], bins=[0, A, B, C, np.inf], labels=['LowAtk', 'Average', 'HighAtk', 'Insane'])

-  then copy it into a new cell and run it

In [63]:
A = 111.750000
B = 146.500000
C = 182.000000

In [64]:
creatures["CombinedAttack"] = pd.cut(creatures["CombinedAttack"], bins=[0, A, B, C, np.inf], labels=['LowAtk', 'Average', 'HighAtk', 'Insane'])
creatures["CombinedAttack"].head(10)

0    Average
1    Average
2    HighAtk
3     Insane
4    Average
5    Average
6     Insane
7     Insane
8     Insane
9     LowAtk
Name: CombinedAttack, dtype: category
Categories (4, object): ['LowAtk' < 'Average' < 'HighAtk' < 'Insane']

# Question 3: Stuck
  - What values did you find for the 25, 50, and 75 percentiles you modified the bin (A, B, and C) with?
  
  - provide the first 10 values in the newly modified ('CombinedAttack') column

1. A = 111.750000, B = 146.500000, C = 182.000000

2. Value 1 = Average, Value 2 = Average, Value 3 = HighAtk, Value 4 = Insane, Value 5 = Average, Value 6 = Average, Value 7 = Insane, Value 8 = Insane, Value 9 = Insane, Value 10 = LowAtk

# Part 3: Data Extraction

# Question 4:
- Find the following (Provide the output for each question and provide the answers in a labeled cell after you find them):

      1) Pokemon with the highest 'Speed'
      2) Pokemon with the lowest 'SpDef'
      3) How many pokemon are in each 'Generation'? (Hint: you've already done something similar in a previous question)

- Example: Say I asked for the lowest speed pokemon
  - You would need to find the index of the pokemon with the lowest speed - then use .loc to return the entire values of that row to see which pokemon (similar to the code below)

        speed_index = creatures['Speed'].idxmin()
        speed_pokemon = creatures.loc[speed_index]
        speed_pokemon


  - use .idxmin() if you need the lowest column value
  - use .idxmax() if you need the largest column value
  - these will only find the first occurence if there are multiple pokemon with the same value - that is fine for this assignment


In [65]:
speed_index = creatures['Speed'].idxmax()
fastest_pokemon = creatures.loc[speed_index]

In [66]:
spdef_index = creatures['SpDef'].idxmin()
lowest_spdef_pokemon = creatures.loc[spdef_index]

In [67]:
generation_counts = creatures['Generation'].value_counts()

In [68]:
print(fastest_pokemon)

Num                             386
Name              DeoxysSpeed Forme
Type1                       Psychic
Type2                           NaN
Total                           600
HP                               50
Attack                           95
Defense                          90
SpAtk                            95
SpDef                            90
Speed                           180
Generation                        3
Legendary                      True
CombinedAttack               Insane
Name: 431, dtype: object


In [69]:
print(lowest_spdef_pokemon)

Num                     10
Name              Caterpie
Type1                  Bug
Type2                  NaN
Total                  195
HP                      45
Attack                  30
Defense                 35
SpAtk                   20
SpDef                   20
Speed                   45
Generation               1
Legendary            False
CombinedAttack      LowAtk
Name: 13, dtype: object


In [70]:
print(generation_counts)

1    166
5    165
3    160
4    121
2    106
6     82
Name: Generation, dtype: int64


1. Deoxys (Speed Forme)

2. Caterpie

3. Gen 1 = 166, Gen 5 = 165, Gen 3 = 160, Gen 4 = 121, Gen 2 = 106, Gen 6 = 82

# Question 5:
  - Access the row for your favorite pokemon and show the output
  - Once done with that - rename the first 3 columns ('Num', 'Name', and 'Type1') to new names (names can be anything you want)
     - show new names once done using .info()
      - Remember:
                df = df.rename(columns={'Num': 'New_Name1', 'Name': 'New_Name2', 'Type1': 'New_Name3'})

In [78]:
creatures.iloc[30]

ID                      25
Nickname           Pikachu
PrimaryType       Electric
Type2                  NaN
Total                  320
HP                      35
Attack                  55
Defense                 40
SpAtk                   50
SpDef                   50
Speed                   90
Generation               1
Legendary            False
CombinedAttack      LowAtk
Name: 30, dtype: object

In [ ]:
creatures = creatures.rename(columns={'Num': 'ID', 'Name': 'Nickname', 'Type1': 'PrimaryType'})

In [72]:
creatures.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 800 entries, 0 to 799
Data columns (total 14 columns):
 #   Column          Non-Null Count  Dtype   
---  ------          --------------  -----   
 0   ID              800 non-null    int64   
 1   Nickname        800 non-null    object  
 2   PrimaryType     800 non-null    object  
 3   Type2           414 non-null    object  
 4   Total           800 non-null    int64   
 5   HP              800 non-null    int64   
 6   Attack          800 non-null    int64   
 7   Defense         800 non-null    int64   
 8   SpAtk           800 non-null    int64   
 9   SpDef           800 non-null    int64   
 10  Speed           800 non-null    int64   
 11  Generation      800 non-null    int64   
 12  Legendary       800 non-null    bool    
 13  CombinedAttack  800 non-null    category
dtypes: bool(1), category(1), int64(9), object(3)
memory usage: 76.9+ KB
